src/sql/workbench/contrib/notebook/browser/cellViews/markdownToolbar.component.ts

```diff
@@ -203,15 +213,41 @@ export class MarkdownToolbarComponent extends AngularDisposable {
 				this._actionBar.setContent(this._taskbarContent);
 			}
 		}
+		this._notebookEditor = this._notebookService.findNotebookEditor(this.cellModel?.notebookModel?.notebookUri);
 	}
 
-	public onInsertButtonClick(event: MouseEvent, type: MarkdownButtonType): void {
-		let go = new MarkdownTextTransformer(this._notebookService, this.cellModel, this._instantiationService);
-		let trigger = event.target as HTMLElement;
-		go.transformText(type, trigger);
+	public async onInsertButtonClick(event: MouseEvent, type: MarkdownButtonType): Promise<void> {
+		event.preventDefault();
+		event.stopPropagation();
+		let triggerElement = event.target as HTMLElement;
+		let needsTransform = true;
+		let calloutResult: ILinkCalloutDialogOptions;
+		if (type === MarkdownButtonType.LINK_PREVIEW) {
+			calloutResult = await this.createCallout(type, triggerElement);
+			// If no URL is present, no-op
+			if (!calloutResult.insertEscapedLinkUrl) {
+				return;
+			}
+			// If cell edit mode isn't WYSIWYG, use result from callout. No need for further transformation.
+			if (this.cellModel.currentMode !== CellEditModes.WYSIWYG) {
+				needsTransform = false;
+			} else {
+				// Otherwise, re-focus on the output element, and insert the link directly.
+				this.output?.nativeElement?.focus();
+				// Callout is responsible for returning escaped strings
+				document.execCommand('insertHTML', false, `<a href="${calloutResult?.insertEscapedLinkUrl}">${calloutResult?.insertEscapedLinkLabel}</a>`);
```

Comment:
Now you're using the markdown-escaped URl and label here which probably isn't correct for inserting into an anchor tag.

This is the kind of thing I was referring to when I suggested letting the callers handle the escaping.

Reply:
Thanks. I thought about this a bit more. I would like the markdown to be used as-is, so I think that it makes sense to return escaped markdown. For the other strings, can return unescaped and leave up to callers to handle as they wish. I've changed the variable names to reflect this.